File: packages/engine/src/difficulty.ts
import type { Analysis, BandConfig, LevelRules } from './types.js';

export function difficultyScore(band: BandConfig, rules: LevelRules, analysis: Analysis): number {
  const operationWeight = Math.max(0, band.allowedOperations.length - 1) * 8;
  const rangeWeight = Math.max(0, band.numberRange[1] - band.numberRange[0]);
  const chainWeight = Math.max(0, band.maxChain - band.minChain) * 4;
  const scarcityWeight = Math.max(0, 6 - Math.min(6, analysis.solutions.length)) * 5;
  const pressureWeight = (rules.timeLimitMs ? 10 : 0) + (rules.moveLimit ? 5 : 0);
  return operationWeight + rangeWeight + chainWeight + scarcityWeight + pressureWeight;
}

File: flutter/packages/game_engine/lib/src/difficulty.dart
import 'package:poko_game_engine/src/types.dart';

int difficultyScore(BandConfig band, LevelRules rules, Analysis analysis) {
  final operationWeight = _max(0, band.allowedOperations.length - 1) * 8;
  final rangeWeight = _max(0, band.numberRange.max - band.numberRange.min);
  final chainWeight = _max(0, band.maxChain - band.minChain) * 4;
  final scarcityWeight = _max(0, 6 - _min(6, analysis.solutions.length)) * 5;
  final pressureWeight =
      (rules.timeLimitMs == null ? 0 : 10) + (rules.moveLimit == null ? 0 : 5);
  return operationWeight +
      rangeWeight +
      chainWeight +
      scarcityWeight +
      pressureWeight;
}

int _max(int left, int right) => left > right ? left : right;
int _min(int left, int right) => left < right ? left : right;
